<a href="https://colab.research.google.com/github/enjileona/Bootcamp-Data-Science-and-Mechine-Learning/blob/main/1_Tugas_Mini_Project_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np

df = pd.read_csv("/content/german_credit_data.csv")

df

Unnamed: 0  Age     Sex  Job Housing Saving accounts Checking account  \
0             0   67    male    2     own             NaN           little   
1             1   22  female    2     own          little         moderate   
2             2   49    male    1     own          little              NaN   
3             3   45    male    2    free          little           little   
4             4   53    male    2    free          little           little   
..          ...  ...     ...  ...     ...             ...              ...   
995         995   31  female    1     own          little              NaN   
996         996   40    male    3     own          little           little   
997         997   38    male    2     own          little              NaN   
998         998   23    male    2    free          little           little   
999         999   27    male    2     own        moderate         moderate   

     Credit amount  Duration              Purpose  Risk  
0             1169         6             radio/TV  good  
1             5951        48             radio/TV   bad  
2             2096        12            education  good  
3             7882        42  furniture/equipment  good  
4             4870        24                  car   bad  
..             ...       ...                  ...   ...  
995           1736        12  furniture/equipment  good  
996           3857        30                  car  good  
997            804        12             radio/TV  good  
998           1845        45             radio/TV   bad  
999           4576        45                  car  good  

[1000 rows x 11 columns]

In [ ]:
print("### Statistik Deskriptif Kolom Numerik ###")
print(df.describe())

### Statistik Deskriptif Kolom Numerik ###
        Unnamed: 0          Age          Job  Credit amount     Duration
count  1000.000000  1000.000000  1000.000000    1000.000000  1000.000000
mean    499.500000    35.546000     1.904000    3271.258000    20.903000
std     288.819436    11.375469     0.653614    2822.736876    12.058814
min       0.000000    19.000000     0.000000     250.000000     4.000000
25%     249.750000    27.000000     2.000000    1365.500000    12.000000
50%     499.500000    33.000000     2.000000    2319.500000    18.000000
75%     749.250000    42.000000     2.000000    3972.250000    24.000000
max     999.000000    75.000000     3.000000   18424.000000    72.000000


# 1. DATA CLEANSING

In [ ]:
# a. Hapus kolom tak penting (Unnamed)
df = df.loc[:, ~df.columns.str.contains('^Unnamed')]

df

Age     Sex  Job Housing Saving accounts Checking account  Credit amount  \
0     67    male    2     own             NaN           little           1169   
1     22  female    2     own          little         moderate           5951   
2     49    male    1     own          little              NaN           2096   
3     45    male    2    free          little           little           7882   
4     53    male    2    free          little           little           4870   
..   ...     ...  ...     ...             ...              ...            ...   
995   31  female    1     own          little              NaN           1736   
996   40    male    3     own          little           little           3857   
997   38    male    2     own          little              NaN            804   
998   23    male    2    free          little           little           1845   
999   27    male    2     own        moderate         moderate           4576   

     Duration              Purpose  Risk  
0           6             radio/TV  good  
1          48             radio/TV   bad  
2          12            education  good  
3          42  furniture/equipment  good  
4          24                  car   bad  
..        ...                  ...   ...  
995        12  furniture/equipment  good  
996        30                  car  good  
997        12             radio/TV  good  
998        45             radio/TV   bad  
999        45                  car  good  

[1000 rows x 10 columns]

In [ ]:
# b. Cek dan hapus duplikasi
print("Jumlah duplikat:", df.duplicated().sum())
df = df.drop_duplicates()

df

Jumlah duplikat: 0


Age     Sex  Job Housing Saving accounts Checking account  Credit amount  \
0     67    male    2     own             NaN           little           1169   
1     22  female    2     own          little         moderate           5951   
2     49    male    1     own          little              NaN           2096   
3     45    male    2    free          little           little           7882   
4     53    male    2    free          little           little           4870   
..   ...     ...  ...     ...             ...              ...            ...   
995   31  female    1     own          little              NaN           1736   
996   40    male    3     own          little           little           3857   
997   38    male    2     own          little              NaN            804   
998   23    male    2    free          little           little           1845   
999   27    male    2     own        moderate         moderate           4576   

     Duration              Purpose  Risk  
0           6             radio/TV  good  
1          48             radio/TV   bad  
2          12            education  good  
3          42  furniture/equipment  good  
4          24                  car   bad  
..        ...                  ...   ...  
995        12  furniture/equipment  good  
996        30                  car  good  
997        12             radio/TV  good  
998        45             radio/TV   bad  
999        45                  car  good  

[1000 rows x 10 columns]

In [ ]:
# c. Menghitung jumlah missing values
df[['Duration']].isnull().sum()


# Mengisi missing dengan nilai 0
# yang mengindikasikan proses belum selesai, sehingga durasinya dianggap 0 untuk saat ini.
df['Duration'] = df['Duration'].fillna(0)
df

Age     Sex  Job Housing Saving accounts Checking account  Credit amount  \
0     67    male    2     own             NaN           little           1169   
1     22  female    2     own          little         moderate           5951   
2     49    male    1     own          little              NaN           2096   
3     45    male    2    free          little           little           7882   
4     53    male    2    free          little           little           4870   
..   ...     ...  ...     ...             ...              ...            ...   
995   31  female    1     own          little              NaN           1736   
996   40    male    3     own          little           little           3857   
997   38    male    2     own          little              NaN            804   
998   23    male    2    free          little           little           1845   
999   27    male    2     own        moderate         moderate           4576   

     Duration              Purpose  Risk  
0           6             radio/TV  good  
1          48             radio/TV   bad  
2          12            education  good  
3          42  furniture/equipment  good  
4          24                  car   bad  
..        ...                  ...   ...  
995        12  furniture/equipment  good  
996        30                  car  good  
997        12             radio/TV  good  
998        45             radio/TV   bad  
999        45                  car  good  

[1000 rows x 10 columns]

In [ ]:
#d. menangani outlier

# Hitung Q1, Q3, dan IQR
Q1          = df['Credit amount'].quantile(0.25)
Q3          = df['Credit amount'].quantile(0.75)
IQR         = Q3 - Q1
batas_atas  = Q3 + 1.5 * IQR
batas_bawah = Q1 - 1.5 * IQR

print(f"Q1: {Q1}, Q3: {Q3}, IQR: {IQR}")
print(f"Batas Atas: {batas_atas}, Batas Bawah: {batas_bawah}")

# Deteksi outlier
outliers = df[(df['Credit amount'] > batas_atas) | (df['Credit amount'] < batas_bawah)]
outliers

# Capping outlier pada kolom 'jumlah_pinjaman'
df['Credit amount'] = df['Credit amount'].clip(
   lower = batas_bawah,
   upper = batas_atas
)

df


Q1: 1365.5, Q3: 3972.25, IQR: 2606.75
Batas Atas: 7882.375, Batas Bawah: -2544.625


Age     Sex  Job Housing Saving accounts Checking account  Credit amount  \
0     67    male    2     own             NaN           little         1169.0   
1     22  female    2     own          little         moderate         5951.0   
2     49    male    1     own          little              NaN         2096.0   
3     45    male    2    free          little           little         7882.0   
4     53    male    2    free          little           little         4870.0   
..   ...     ...  ...     ...             ...              ...            ...   
995   31  female    1     own          little              NaN         1736.0   
996   40    male    3     own          little           little         3857.0   
997   38    male    2     own          little              NaN          804.0   
998   23    male    2    free          little           little         1845.0   
999   27    male    2     own        moderate         moderate         4576.0   

     Duration              Purpose  Risk  
0           6             radio/TV  good  
1          48             radio/TV   bad  
2          12            education  good  
3          42  furniture/equipment  good  
4          24                  car   bad  
..        ...                  ...   ...  
995        12  furniture/equipment  good  
996        30                  car  good  
997        12             radio/TV  good  
998        45             radio/TV   bad  
999        45                  car  good  

[1000 rows x 10 columns]

# 2. Transformasi Data dan Rekayasa Fitur Sederhana

In [ ]:
# a. Membuat fitur 'kategori_Age' dengan binning
# 'umur' menjadi 3 kategori: Muda, Dewasa, Lansia
bins_Age                      = [0, 30, 60, np.inf]
labels_Age                    = ['Young', 'Adult', 'Elderly']

df['Age_category'] = pd.cut(df['Age'], bins=bins_Age, labels=labels_Age)

df

Age     Sex  Job Housing Saving accounts Checking account  Credit amount  \
0     67    male    2     own             NaN           little         1169.0   
1     22  female    2     own          little         moderate         5951.0   
2     49    male    1     own          little              NaN         2096.0   
3     45    male    2    free          little           little         7882.0   
4     53    male    2    free          little           little         4870.0   
..   ...     ...  ...     ...             ...              ...            ...   
995   31  female    1     own          little              NaN         1736.0   
996   40    male    3     own          little           little         3857.0   
997   38    male    2     own          little              NaN          804.0   
998   23    male    2    free          little           little         1845.0   
999   27    male    2     own        moderate         moderate         4576.0   

     Duration              Purpose  Risk Age_category  
0           6             radio/TV  good      Elderly  
1          48             radio/TV   bad        Young  
2          12            education  good        Adult  
3          42  furniture/equipment  good        Adult  
4          24                  car   bad        Adult  
..        ...                  ...   ...          ...  
995        12  furniture/equipment  good        Adult  
996        30                  car  good        Adult  
997        12             radio/TV  good        Adult  
998        45             radio/TV   bad        Young  
999        45                  car  good        Young  

[1000 rows x 11 columns]

In [ ]:
# b1. membuat fitur 'Duration_month'
# Kita gunakan kolom 'Duration' yang berisi hari, kemudian dibagi 30
# dan isi NaN dengan 0.

df['Duration_month'] = (df['Duration'].fillna(0) / 30)

df

Age     Sex  Job Housing Saving accounts Checking account  Credit amount  \
0     67    male    2     own             NaN           little         1169.0   
1     22  female    2     own          little         moderate         5951.0   
2     49    male    1     own          little              NaN         2096.0   
3     45    male    2    free          little           little         7882.0   
4     53    male    2    free          little           little         4870.0   
..   ...     ...  ...     ...             ...              ...            ...   
995   31  female    1     own          little              NaN         1736.0   
996   40    male    3     own          little           little         3857.0   
997   38    male    2     own          little              NaN          804.0   
998   23    male    2    free          little           little         1845.0   
999   27    male    2     own        moderate         moderate         4576.0   

     Duration              Purpose  Risk Age_category  Duration_month  
0           6             radio/TV  good      Elderly             0.2  
1          48             radio/TV   bad        Young             1.6  
2          12            education  good        Adult             0.4  
3          42  furniture/equipment  good        Adult             1.4  
4          24                  car   bad        Adult             0.8  
..        ...                  ...   ...          ...             ...  
995        12  furniture/equipment  good        Adult             0.4  
996        30                  car  good        Adult             1.0  
997        12             radio/TV  good        Adult             0.4  
998        45             radio/TV   bad        Young             1.5  
999        45                  car  good        Young             1.5  

[1000 rows x 12 columns]

In [ ]:
# b2. membuat fitur 'Duration_category'

df['Duration_category'] = pd.cut(
    df['Duration_month'],
    bins=[-1, 1, 2, df['Duration_month'].max()],
    labels=['Short', 'Medium', 'Long']
)

df

Age     Sex  Job Housing Saving accounts Checking account  Credit amount  \
0     67    male    2     own             NaN           little         1169.0   
1     22  female    2     own          little         moderate         5951.0   
2     49    male    1     own          little              NaN         2096.0   
3     45    male    2    free          little           little         7882.0   
4     53    male    2    free          little           little         4870.0   
..   ...     ...  ...     ...             ...              ...            ...   
995   31  female    1     own          little              NaN         1736.0   
996   40    male    3     own          little           little         3857.0   
997   38    male    2     own          little              NaN          804.0   
998   23    male    2    free          little           little         1845.0   
999   27    male    2     own        moderate         moderate         4576.0   

     Duration              Purpose  Risk Age_category  Duration_month  \
0           6             radio/TV  good      Elderly             0.2   
1          48             radio/TV   bad        Young             1.6   
2          12            education  good        Adult             0.4   
3          42  furniture/equipment  good        Adult             1.4   
4          24                  car   bad        Adult             0.8   
..        ...                  ...   ...          ...             ...   
995        12  furniture/equipment  good        Adult             0.4   
996        30                  car  good        Adult             1.0   
997        12             radio/TV  good        Adult             0.4   
998        45             radio/TV   bad        Young             1.5   
999        45                  car  good        Young             1.5   

    Duration_category  
0               Short  
1              Medium  
2               Short  
3              Medium  
4               Short  
..                ...  
995             Short  
996             Short  
997             Short  
998            Medium  
999            Medium  

[1000 rows x 13 columns]

In [ ]:
# c. Terapkan Ordinal Encoding pada kolom 'Saving Account'

mapping_Saving_accounts = {'little': 1, 'moderate': 2, 'quite rich': 3, 'rich': 4}

df['Saving_accounts_oe'] = df['Saving accounts'].map(mapping_Saving_accounts)
df

Age     Sex  Job Housing Saving accounts Checking account  Credit amount  \
0     67    male    2     own             NaN           little         1169.0   
1     22  female    2     own          little         moderate         5951.0   
2     49    male    1     own          little              NaN         2096.0   
3     45    male    2    free          little           little         7882.0   
4     53    male    2    free          little           little         4870.0   
..   ...     ...  ...     ...             ...              ...            ...   
995   31  female    1     own          little              NaN         1736.0   
996   40    male    3     own          little           little         3857.0   
997   38    male    2     own          little              NaN          804.0   
998   23    male    2    free          little           little         1845.0   
999   27    male    2     own        moderate         moderate         4576.0   

     Duration              Purpose  Risk Age_category  Duration_month  \
0           6             radio/TV  good      Elderly             0.2   
1          48             radio/TV   bad        Young             1.6   
2          12            education  good        Adult             0.4   
3          42  furniture/equipment  good        Adult             1.4   
4          24                  car   bad        Adult             0.8   
..        ...                  ...   ...          ...             ...   
995        12  furniture/equipment  good        Adult             0.4   
996        30                  car  good        Adult             1.0   
997        12             radio/TV  good        Adult             0.4   
998        45             radio/TV   bad        Young             1.5   
999        45                  car  good        Young             1.5   

    Duration_category  Saving_accounts_oe  
0               Short                 NaN  
1              Medium                 1.0  
2               Short                 1.0  
3              Medium                 1.0  
4               Short                 1.0  
..                ...                 ...  
995             Short                 1.0  
996             Short                 1.0  
997             Short                 1.0  
998            Medium                 1.0  
999            Medium                 2.0  

[1000 rows x 14 columns]

# 3. EDA

Define Your Question

1. Bagaimana hubungan kategori usia dengan risk credit?
* Variabel 1: Age_category
* variabel 2: Risk

2. Apakah durasi pinjaman yang lebih panjang berkaitan dengan peningkatan risiko gagal bayar?
* Variabel 1: Duration_category
* Variabel 2: Risk

UNIVARIATE ANALYSIS

In [ ]:
# a. Menampilkan Age_category

print("### Statistik Deskriptif Kolom 'Age_category' ###")
print(df ['Age_category'].value_counts())

### Statistik Deskriptif Kolom 'Age_category' ###
Age_category
Adult      544
Young      411
Elderly     45
Name: count, dtype: int64


Berdasarkan hasil tersebut, didapati bahwa pengguna merupakan category usia Dewasa atau 30 - 60 tahun, yang artinya pengguna sudah berada pada usia produktif. Disusul dengan pengguna pada usia young atau kurang dari 30 tahun.

In [ ]:
# b. Menampilkan Duration_category

print("### Statistik Deskriptif Kolom 'Duration_category' ###")
print(df ['Duration_category'].value_counts())

### Statistik Deskriptif Kolom 'Duration_category' ###
Duration_category
Short     827
Medium    172
Long        1
Name: count, dtype: int64


Berdasarkan hasil tersebut, didapati bahwa pengguna kredit terbanyak berada pada durasi Short atau 0 - 1 bulan. Sedangkan paling sedikit adalah durasi long atau lebih dari 2 bulan.

MULTIVARIATE ANALYSIS

In [ ]:
# a. Menampilkan hubungan antara Age_categori dan risk

crosstab_Age_risk = pd.crosstab(df['Age_category'], df['Risk'])
print("### Cross Tabulation: Age_Category vs Risk ###")
print(crosstab_Age_risk)

### Cross Tabulation: Age_Category vs Risk ###
Risk          bad  good
Age_category           
Young         148   263
Adult         142   402
Elderly        10    35


In [ ]:
# b. Menampilkan hubungan antara Duration dan risk

df[df["Duration_month"] < 1]['Risk'].value_counts()

Risk
good    583
bad     204
Name: count, dtype: int64

#Conclutions

**1. Hubungan antara Age_category dan Risk**:
Berdasarkan data, terlihat bahwa nasabah berusia muda (Young) memiliki risiko gagal bayar lebih tinggi dibanding kelompok usia lainnya, dengan sekitar sepertiga dari total nasabah muda masuk kategori bad risk. Sementara itu, kelompok dewasa (Adult) menunjukkan stabilitas terbaik, karena mayoritas besar nasabah dewasa berada pada kategori good risk. Untuk kelompok lansia (Elderly), meskipun jumlahnya relatif sedikit, proporsi risiko gagal bayar lebih rendah dibandingkan kelompok muda, sehingga secara umum nasabah lansia cenderung lebih aman.





**2. Hubungan antara Duration Month dan Risk**:
Berdasarkan data untuk nasabah dengan durasi pinjaman kurang dari 1 bulan, terlihat bahwa mayoritas berada pada kategori good risk, yaitu sebanyak 583 nasabah, sedangkan yang masuk kategori bad risk sebanyak 204 nasabah. Hal ini menunjukkan bahwa meskipun ada risiko gagal bayar pada kelompok durasi singkat, proporsi nasabah yang mampu membayar kewajiban tetap lebih tinggi. Dengan kata lain, nasabah dengan durasi pinjaman di bawah 1 bulan cenderung memiliki profil risiko yang lebih aman, karena sekitar tiga per empat dari mereka tergolong good risk.